### PR TITLE
Major revamp of install.md doc

### PR DESCRIPTION
**Pull request by beamcodeup**
Regarding Issue #132

The installation doc deserves a major revamp for the sake of legiblity and clear instructions. The formatting and structures have been updated. New steps are added to address missing dependencies when building and installing in the virutalenv and GCP service setup (API, service account, etc).

### CODE REVIEW

File: docs/install.md [hunk 1]
```diff
@@ -1,77 +1,197 @@
...
+directory so that when Evidence is added back into Turbinia that the other
+worker nodes can process it. Turbinia can also write output to GCS even when
+running locally (set the `GCS_OUTPUT_PATH` variable in the config).
+
+## Google Cloud Platform (GCP) Setup
+
+Turbinia can read Evidence from either cloud Persistent Disks, or from GCS
+objects. Turbinia can also write output to GCS (set the `GCS_OUTPUT_PATH`
+variable in the config). Note that you can operate multiple Turbinia instances
+within the same GCP Project as long as your careful to make sure your config
```
Comment (aarontp) [inline docs/install.md]: s/your careful/you're careful/

File: docs/install.md [hunk 2]
```diff
@@ -1,77 +1,197 @@
...
+
+If you have raw disk images from physical machines or otherwise that you want to
+process in the cloud, you can process them directly from GCS. This can be used
+when the images do not originate from VMs/containers that are running in the
+cloud. You can potentially convert raw images to Persistent Disks, but that
+requires processing the image first, and so using GCS is recommended in this
+case. Processing Evidence directly from GCS requires that you use [GCS
+FUSE](https://cloud.google.com/storage/docs/gcs-fuse) and mount your bucket at a
+common path on each worker node. Once your GCS bucket is mounted, you can
+process these images as 'rawdisk' Evidence.
```
Comment (aarontp) [inline docs/install.md]: So, this section was true when it was written, but we've since found that processing images directly from GCS does not work well.  Feel free to leave it for now, but we should make sure to update it to show how to use GoogleCloudDiskRawEmbedded disks instead.

File: docs/install.md [hunk 3]
```diff
@@ -1,77 +1,197 @@
...
+    *   Create a new Pub/Sub topic and subscription **(pull type)**
+    *   Please take a note of the topic and subscription name for the
+        configuration steps
+*   Enable [Cloud
+    Datastore](https://console.cloud.google.com/apis/api/datastore.googleapis.com)
+    *   Go to Datastore in the cloud console
+    *   Hit the `Create Entity` button
+    *   Select the region that your data should be in. No need to create any
+        Entities after selecting your region
+*   Enable [Cloud Deployment Manager
```
Comment (aarontp) [inline docs/install.md]: What's this for?

File: docs/install.md [hunk 4]
```diff
@@ -1,77 +1,197 @@
...
+*   Create a new GCE server VM from a recent version of Debian or Ubuntu
+    *   This should work on other Linux flavors, but these are untested. Feel
+        free to test and fix them ;)
+*   Follow [Core Installation Steps](#core-installation-steps)
+*   Create a new image from the server VM's disk
+*   Create a new Instance Template using the newly created image
+*   Create a new Managed Instance Group from the newly created Instance Template
+*   Create a new GCS bucket and create new directories for `scripts` and
+    `output/logs`
+*   Mount your GCS bucket on your server VM
```
Comment (aarontp) [inline docs/install.md]: Might want to add something like "using GCS FUSE" so it's clear how it gets mounted.

File: docs/install.md [hunk 5]
```diff
@@ -1,77 +1,197 @@
...
+*   cp `turbinia/tools/gcp_init/*.sh` into your locally mounted copy of
+    `$your_bucket/scripts`
+*   Edit the variables in `scripts/start-wrapper.sh` and
+    `scripts/start-turbinia-common.sh` as appropriate
+*   ***NOTE:*** Please note that the `start-wrapper.sh` script has a
```
Comment (aarontp) [inline docs/install.md]: Maybe s/Please notes that the/The/ since 'note' is here twice?

File: docs/install.md [hunk 6]
```diff
@@ -1,77 +1,197 @@
...
+    `GOOGLE_APPLICATION_CREDENTIALS` environment var in the middle of the script
+    that needs to be updated
+*   In your worker VM, add a new custom metadata key `startup-script-url`
```
Comment (aarontp) [inline docs/install.md]: This actually needs to be done when the "Instance Template" is created above.  Maybe make this a bullet point under that?

File: docs/install.md [hunk 7]
```diff
@@ -1,77 +1,197 @@
...
+    pointing to `gs://$your_bucket/scripts/start-wrapper.sh`
+*   Upon start, your VM should mount your GCS Bucket, and copy the start scripts
+    into the home directory of the Turbinia user and will then start the
+    Turbinia worker
+*   If you are running in GCP, you may also want to install [GCS
```
Comment (aarontp) [inline docs/install.md]: This section already assumes that we're running in GCP, so I think we can remove the first part up to the comma.  Also, this is assumed to be installed at the steps above when referencing the scripts, so maybe move this to above the first mention of "GCS Bucket" in this section?

We could also add the note here with something like: "We're using GCS FUSE here for convenience to keep scripts and log files, but this isn't strictly necessary for Turbinia to run if you have an alternate means to save logging data".

File: docs/install.md [hunk 8]
```diff
@@ -1,77 +1,197 @@
...
+    FUSE](https://cloud.google.com/storage/docs/gcs-fuse)
+
+## Core Installation Steps
+
+*   Install python-dev, build essential, pip and setuptools
+    *   `sudo apt-get install python-dev build-essential python-setuptools
+        python-pip`
+*   Install virtualenv and git
+    *   `sudo apt-get install python-virtualenv git`
```
Comment (aarontp) [inline docs/install.md]: Can we combine these into the previous apt-get install line?

File: docs/install.md [hunk 9]
```diff
@@ -1,77 +1,197 @@
...
+*   Create (once) and activate Virtualenv
+    *   `virtualenv turbinia-env && . turbinia-env/bin/activate`
+    *   ***NOTE:*** the next time you need to use virtualenv, just log in as
+        turbinia and activiate virtualenv without recreating it
+*   Do not exit the Virtualenv until you have completed all the steps!
+*   Continue to [Inside the Virtualenv](#inside-the-virtualenv)
+
+### Inside the Virtualenv
+
+#### Google Cloud SDK, IAM roles, auth credentials
```
Comment (aarontp) [inline docs/install.md]: This section doesn't actually have to be done under the virtual env.  Maybe we can move it out?

File: docs/install.md [hunk 10]
```diff
@@ -1,77 +1,197 @@
...
+*   Alternately you can run Turbinia under your own credentials **(NOT
+    RECOMMENDED)**
+    *   Run `gcloud auth login` (may require you to copy/paste url to browser)
+    *   Or run `gcloud auth application-default login`
+*   Continue to [Build and Configure](#build-and-configure)
+
+#### Build and Configure
+
+*   Install Turbinia
+    *   `sudo apt-get install liblzma-dev`
```
Comment (aarontp) [inline docs/install.md]: Can we also add this package to the previous 'apt-get install' line?

File: docs/install.md [hunk 11]
```diff
@@ -1,77 +1,197 @@
...
-
-#### Persistent Disks
-Persistent disks should be the default when processing disks that come from the Cloud.  The account you run Turbinia as must have access to the Persistent Disks that you want to process.  If you add a GoogleCloudDisk Evidence type, the worker node will attach the disk automatically before it runs its tasks.
-
-#### GCS objects
-If you have raw disk images from physical machines or otherwise that you want to process in the cloud, you can process them directly from GCS.  This can be used when the images do not originate from VMs/containers that are running in the cloud.  You can potentially convert raw images to Persistent Disks, but that requires processing the image first, and so using GCS is recommended in this case.  Processing Evidence directly from GCS requires that you use [GCS FUSE](https://cloud.google.com/storage/docs/gcs-fuse) and mount your bucket at a common path on each worker node.  Once your GCS bucket is mounted, you can process these images as 'rawdisk' Evidence.
+# High-Level Setup
+
+Turbinia can be run either in the Google Cloud, or on local machines. If you run
```
Comment (aarontp) [inline docs/install.md]: General comment for the doc:  I don't know much about markdown style, but is manually adding newlines into the text the idiomatic way to write these files?  There are line breaks in some strange places (e.g. in the middle of bash one-liners), but I guess it doesn't change their readability.

File: docs/install.md [hunk 12]
```diff
@@ -1,77 +1,197 @@
...
+requires processing the image first, and so using GCS is recommended in this
+case. Processing Evidence directly from GCS requires that you use [GCS
+FUSE](https://cloud.google.com/storage/docs/gcs-fuse) and mount your bucket at a
+common path on each worker node. Once your GCS bucket is mounted, you can
+process these images as 'rawdisk' Evidence.
+
+# Instructions
+
+## GCP Project (Cloud Pub/Sub, Cloud Function, Cloud Datastore Setup)
+
```
Comment (aarontp) [inline docs/install.md]: You mention it above, but maybe add a line here that this section is required when running Turbinia on GCP or on local machines.

File: docs/install.md [hunk 13]
```diff
@@ -1,77 +1,197 @@
...
+    Datastore](https://console.cloud.google.com/apis/api/datastore.googleapis.com)
+    *   Go to Datastore in the cloud console
+    *   Hit the `Create Entity` button
+    *   Select the region that your data should be in. No need to create any
+        Entities after selecting your region
+*   Enable [Cloud Deployment Manager
+    V2](https://console.cloud.google.com/apis/library/deploymentmanager.googleapis.com)
+    for `gcloud`
+
+## Local Turbinia
```
Comment (aarontp) [inline docs/install.md]: Maybe just move the "local turbinia" section below GCP Turbinia since it's just a pointer to Core?

File: docs/install.md [hunk 14]
```diff
@@ -1,77 +1,197 @@
...
+        *   `sudo mkdir /etc/turbinia`
+        *   `cp <localgitpath>/turbinia/config/turbinia_config.py /etc/turbinia/`
+    *   `/home/turbinia/.turbinia`
+        *   `cp <localgitpath>/turbinia/config/turbinia_config.py /home/turbinia/.turbiniarc`
+    *   Directly configure `<localgitpath>/turbinia/config/turbinia_config.py`
+    *   ***NOTE***: Match the `PUBSUB_TOPIC` variable in the configuration to the name
+        of the topic and subscription you created in the GCP.
+*   Continue to [Deploy the Cloud Functions](#deploy-the-cloud-functions)
+
+### Deploy the Cloud Functions
```
Comment (aarontp) [inline docs/install.md]: Maybe just inline this section with the section that references it since it's so short?

File: docs/install.md [hunk 15]
```diff
@@ -1,77 +1,197 @@
...
+    *   `curl -o plaso-requirements.txt
+        https://raw.githubusercontent.com/log2timeline/plaso/master/requirements.txt`
+    *   `pip install -r plaso-requirements.txt`
+*   Update Python Path
+    *   Until [Issue #39](https://github.com/google/turbinia/issues/39) is
+        fixed, we need to adjust the `$PYTHONPATH` from the root of the Turbinia
+        repository:
+        *   ``cd turbinia; export PYTHONPATH=$PYTHONPATH:`pwd`; cd -``
+    *   And set it in your .bashrc file:
+        *   ``cd turbinia; echo "export PYTHONPATH=$PYTHONPATH:`pwd`" >>
```
Comment (aarontp) [inline docs/install.md]: Actually, I just realized that this will add the Turbinia directory a second time since it's already added to the $PYTHONPATH above.

You might want to  test it, but I think you can change this to:
``echo "export PYTHONPATH=$PYTHONPATH" >> ~/.bashrc``

File: docs/install.md (hunk 1 above)
Comment (beamcodeup) [reply docs/install.md]: Done.

File: docs/install.md (hunk 11 above)
Comment (beamcodeup) [reply docs/install.md]: WAI - used md linter.

File: docs/install.md (hunk 2 above)
Comment (beamcodeup) [reply docs/install.md]: Added a todo line

File: docs/install.md (hunk 12 above)
Comment (beamcodeup) [reply docs/install.md]: Done.

File: docs/install.md (hunk 3 above)
Comment (beamcodeup) [reply docs/install.md]: This allows us to manage our GCP via the tool `gcloud`.

File: docs/install.md (hunk 13 above)
Comment (beamcodeup) [reply docs/install.md]: I'll leave this as-is because it's clear and easy to follow.

File: docs/install.md (hunk 4 above)
Comment (beamcodeup) [reply docs/install.md]: Done

File: docs/install.md (hunk 5 above)
Comment (beamcodeup) [reply docs/install.md]: Done

File: docs/install.md (hunk 6 above)
Comment (beamcodeup) [reply docs/install.md]: I realized this section needs to be re-worked abit. I broke this section into smaller sub sections.

File: docs/install.md (hunk 8 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 9 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 10 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 14 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 15 above)
Comment (beamcodeup) [reply docs/install.md]: tested and done.

File: docs/install.md (hunk 7 above)
Comment (beamcodeup) [reply docs/install.md]: Done

**Comment by beamcodeup** [comment]
@aarontp PTAL.

File: docs/install.md [hunk 16]
```diff
@@ -1,77 +1,220 @@
...
+    Steps](#core-installation-steps)
+*   Otherwise, follow [GCP Turbinia](#gcp-turbinia)
+
+## GCP Turbinia
+
+The following is a one possible configuration and setup for Turbinia in GCP.
+This is still a rough process and future versions will be containerized.
+
+**NOTE:** GCP Turbinia requires two Google Cloud Engine (GCE) instance types:
+1 server and 1 or more workers
```
Comment (aarontp) [inline docs/install.md]: I would probably say:
"When running Turbinia on GCP, it's recommended that you have at least two VM instances (one for the server, and one or more for the workers).  In a small setup, you can also run both the server and worker on a single instance"

File: docs/install.md [hunk 17]
```diff
@@ -1,77 +1,220 @@
...
+*   Create a new Managed Instance Group from the newly created Instance Template
+*   In your worker VM, add a new custom metadata key `startup-script-url`
+    pointing to `gs://$your_bucket/scripts/start-wrapper.sh`
+*   TODO(beamcodeup): Update this section when Issue #132 is resolved.
+*   Upon start, your VM should mount your GCS Bucket, and copy the start scripts
+    into the home directory of the Turbinia user and will then start the
+    Turbinia worker
+*   ***NOTE:*** If the GCS FUSE is used for log files, update the Turbinia config
+    file with a new `LOG_FILE` path to avoid writing to the server logs.
+
```
Comment (aarontp) [inline docs/install.md]: Not sure what is meant by "to avoid writing to the server logs" here.   Maybe you mean something like "to write server logs directly into GCS"?

File: docs/install.md [hunk 18]
```diff
@@ -1,77 +1,220 @@
...
+        GOOGLE_APPLICATION_CREDENTIALS="/home/turbinia/turbinia-service-account-creds.json"`
+*   Add the service account to the gcloud auth **(RECOMMENDED)**
+    *   `gcloud auth list`
+    *   `gcloud auth activate-service-account
+        --key-file=$GOOGLE_APPLICATION_CREDENTIALS`
+*   Alternately you can run Turbinia under your own credentials **(NOT
+    RECOMMENDED)**
+    *   Run `gcloud auth login` (may require you to copy/paste url to browser)
+    *   Or run `gcloud auth application-default login`
+*   Create (once) and activate Virtualenv
```
Comment (aarontp) [inline docs/install.md]: Maybe just move the virtualenv related instructions to the "Inside the virtualenv" section?

File: docs/install.md [hunk 19]
```diff
@@ -1,77 +1,220 @@
...
+        fixed, we need to adjust the `$PYTHONPATH` from the root of the Turbinia
+        repository:
+        *   ``cd turbinia; export PYTHONPATH=$PYTHONPATH:`pwd`; cd -``
+    *   And set it in your .bashrc file:
+        *   ``cd turbinia; echo "export PYTHONPATH=$PYTHONPATH" >> ~/.bashrc ; cd -``
+*   Choose one of the locations for storing the Turbinia config and configure
+    from there:
+    *   `/etc/turbinia/` **(RECOMMENDED)**
+        *   `sudo mkdir /etc/turbinia`
+        *   `cp <localgitpath>/turbinia/config/turbinia_config.py /etc/turbinia/`
```
Comment (aarontp) [inline docs/install.md]: I think for /etc/turbinia we should recommend the filename to be /etc/turbinia/turbinia.conf

File: turbinia/config/turbinia_config.py [hunk 20]
```diff
@@ -1,66 +1,104 @@
...
+# distributed under the License is distributed on an 'AS IS' BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
 from __future__ import unicode_literals
 
 
-"""Dummy Turbinia config file."""
+"""Turbinia Config Template"""
```
Comment (aarontp) [inline turbinia/config/turbinia_config.py]: Are these changes intended to be part of this PR?  Or is this for the init scripts?

File: docs/install.md [hunk 21]
```diff
@@ -1,77 +1,220 @@
...
+case. Processing Evidence directly from GCS requires that you use [GCS
+FUSE](https://cloud.google.com/storage/docs/gcs-fuse) and mount your bucket at a
+common path on each worker node. Once your GCS bucket is mounted, you can
+process these images as 'rawdisk' Evidence.
+
+TODO(beamcodeup): Document how to use GoogleCloudDiskRawEmbedded.
+
+# Instructions
+
+## GCP Project (Cloud Pub/Sub, Cloud Function, Cloud Datastore Setup)
```
Comment (aarontp) [inline docs/install.md]: Should 'Setup' be outside of the parens here?

File: docs/install.md (hunk 16 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 17 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md [hunk 22]
```diff
@@ -1,77 +1,220 @@
...
+        configuration steps
+*   Enable [Cloud
+    Datastore](https://console.cloud.google.com/apis/api/datastore.googleapis.com)
+    *   Go to Datastore in the cloud console
+    *   Hit the `Create Entity` button
+    *   Select the region that your data should be in. No need to create any
+        Entities after selecting your region
+*   Enable [Cloud Deployment Manager
+    V2](https://console.cloud.google.com/apis/library/deploymentmanager.googleapis.com)
+    for managing the GCP via the CLI tool `gclou`
```
[the remark this answers is not in the document]
Comment (beamcodeup) [reply docs/install.md]: done

File: turbinia/config/turbinia_config.py (hunk 20 above)
Comment (beamcodeup) [reply turbinia/config/turbinia_config.py]: I think I adversely edited this file when using sed. Reverting this.

File: docs/install.md (hunk 18 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 21 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 19 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md [hunk 23]
```diff
@@ -1,77 +1,224 @@
...
+    *   `/home/turbinia/.turbinia`
+        *   `cp <localgitpath>/turbinia/config/turbinia_config.py /home/turbinia/.turbiniarc`
+    *   Directly configure `<localgitpath>/turbinia/config/turbinia_config.py`
+    *   ***NOTE***: Match the `PUBSUB_TOPIC` variable in the configuration to the name
+        of the topic and subscription you created in the GCP.
+*   Continue to [Deploy the Cloud Functions](#deploy-the-cloud-functions)
+
+#### Deploy the Cloud Functions
+
+*   `cd <localgitpath>/turbinia/tools/gcf_init && ./deploy_gcf.py`
```
Comment (beamcodeup) [inline docs/install.md]: I tested the installation instruction by following it exactly.

I didn't realize deploy_gcf.py ran gcloud functions deploy with --stage-bucket, which requires GCS. Also, it requires all the configuration varables to be set prior to moving forward. Another thing is if the config needs to be fully set up, I will have to reorder the doc a bit so that GCS is created first before making changes to the config file.

File: docs/install.md (hunk 6 above)
Comment (aarontp) [reply docs/install.md]: Looks better, thanks.

File: docs/install.md (hunk 3 above)
Comment (beamcodeup) [reply docs/install.md]: Originally, gcloud complained that it didn't have access to the deployment manager API. I don't think we need this. Omitted.

File: docs/install.md [hunk 24]
```diff
@@ -1,77 +1,224 @@
...
+
+**NOTE:** This section is mandatory for Turbinia running on GCP or local
+machines.
+
+*   Create or select a Google Cloud Platform project on the [Google Developers
+    Console](https://console.developers.google.com)
+*   Enable [Cloud
+    Functions](https://console.cloud.google.com/apis/library/cloudfunctions.googleapis.com)
+*   Follow the
+    [instruction](https://cloud.google.com/pubsub/docs/quickstart-console) to:
```
[the remark this answers is not in the document]
Comment (beamcodeup) [reply docs/install.md]: Done

File: docs/install.md (hunk 2 above)
Comment (beamcodeup) [reply docs/install.md]: done

File: docs/install.md (hunk 23 above)
Comment (beamcodeup) [reply docs/install.md]: The GCS instruction is already included. So, we are good here.

**Comment by aarontp** [comment]
I think there are still some changes to the config file that you're not intending to make.  :/.    AFAIK, this PR shouldn't have any config changes in it, but the diff shows that there are some, and at least one of them (the BUCKET_NAME change) we don't want.

**Comment by beamcodeup** [comment]
#23
